device/components/RTA_OnDeviceComponentTask.brs

```diff
@@ -39,35 +31,107 @@ sub runTaskThread()
 
 	address = createObject("roSocketAddress")
 	address.setPort(9000)
 
-	udpSocket = createObject("roDatagramSocket")
-	udpSocketId = stri(udpSocket.getID())
-	udpSocket.setMessagePort(m.port)
-	udpSocket.setAddress(address)
-	udpSocket.notifyReadable(true)
+	listenSocket = createObject("roStreamSocket")
+	listenSocketId = listenSocket.getID().toStr()
+	listenSocket.setMessagePort(m.port)
+	listenSocket.setAddress(address)
+	listenSocket.notifyReadable(true)
+	listenSocket.listen(4)
+	m.clientSockets = {}
+
+	receivingRequests = {}
 	m.activeRequests = {}
 
 	while true
 		' If you want to waste three days debugging set this back to 0 :|
-		message = wait(1, m.port)
+		message = wait(1000, m.port)
 		if message <> Invalid then
 			messageType = type(message)
 			if messageType = "roSocketEvent" then
-				messageSocketId = stri(message.getSocketID())
-				if messageSocketId = udpSocketId
-					if udpSocket.isReadable() then
-						receivedString = udpSocket.receiveStr(udpSocket.getCountRcvBuf())
-						verifyAndHandleRequest(receivedString, udpSocket)
+				messageSocketId = message.getSocketID().toStr()
+				if messageSocketId = listenSocketId then
+					if listenSocket.isReadable() then
+						clientSocket = listenSocket.accept()
+						if clientSocket = Invalid then
+							logError("Connection accept failed")
+						else
+							clientSocket.notifyReadable(true)
+							m.clientSockets[clientSocket.getID().toStr()] = clientSocket
+						end if
 					end if
 				else
-					logWarn("Received roSocketEvent for unknown socket")
+					clientSocket = m.clientSockets[messageSocketId]
+					bufferLength = clientSocket.getCountRcvBuf()
+					if bufferLength > 0 then
+						receivingRequest = receivingRequests[messageSocketId]
+						if receivingRequest <> invalid then
+							' This is an existing request and we're now receiving more info for it
+							if receiveDataForRequest(receivingRequest, bufferLength) then
+								receivingRequests.delete(messageSocketId)
+								verifyAndHandleRequest(receivingRequest)
+							end if
+						else
+							ba = createObject("roByteArray")
+							ba[m.requestHeaderSize - 1] = 0
+
+							' Read our request header to know how big our request is
+							clientSocket.receive(ba, 0, m.requestHeaderSize)
+							bufferLength -= m.requestHeaderSize
+							receivingRequest = {
+								"stringLength": unpackInt32LE(ba, 0)
+								"binaryLength": unpackInt32LE(ba, 4)
+								"stringPayload": ""
+								"binaryPayload": createObject("roByteArray")
+								"socket": clientSocket
+							}
+							receivingRequests[messageSocketId] = receivingRequest
+
+							if bufferLength > 0 AND receiveDataForRequest(receivingRequest, bufferLength) then
+								receivingRequests.delete(messageSocketId)
+								verifyAndHandleRequest(receivingRequest)
+							end if
+						end if
+					else
+						logInfo("Client closed connection")
+						clientSocket.close()
+						m.clientSockets.delete(messageSocketId)
+					end if
```

Comment:
This whole section is fairly complicated. It might be better to break it up into at some smaller functions. There's a 6-level and 5-level deep if statement. Very hard to keep track of what's happening.

Reply:
fair point. Can try and find a middle ground on this

Reply:
updated